Failure connecting to SSH agent from Bitwarden

Having configured the Bitwarden SSH Agent in `~/.ssh/config` as

```
IdentityAgent ~/.bitwarden-ssh-agent.sock
```

will cause a connection attempt to hang and/or fail with the stacktrace when the agent is stopped by quitting Bitwarden.app

```
2025-03-15 16:40:19,970 [Thread-184] WARN  ch.cyberduck.core.threading.BackgroundCallable - Failure running background task WorkerBackgroundAction{worker=MountWorker{cache=ch.cyberduck.core.ReverseLookupCache@76e3b45b}}
java.lang.Exception: null
        at ch.cyberduck.core.threading.BackgroundCallable.<init>(BackgroundCallable.java:38) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.threading.DefaultBackgroundExecutor.execute(DefaultBackgroundExecutor.java:71) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.AbstractController.background(AbstractController.java:72) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.ui.cocoa.controller.BrowserController$34.run(BrowserController.java:3308) ~[osx-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.ui.cocoa.controller.BrowserController$37.run(BrowserController.java:3437) ~[osx-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.ui.cocoa.controller.BrowserController$38.cleanup(BrowserController.java:3456) ~[osx-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.threading.AbstractBackgroundAction.cleanup(AbstractBackgroundAction.java:111) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.threading.SessionBackgroundAction.cleanup(SessionBackgroundAction.java:150) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.threading.BackgroundCallable$1.run(BackgroundCallable.java:64) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at org.rococoa.internal.MainThreadUtils$3.callback(MainThreadUtils.java:91) ~[rococoa-core-0.9.1.jar:?]
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source) ~[?:?]
        at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[?:?]
        at com.sun.jna.CallbackReference$DefaultCallbackProxy.invokeCallback(CallbackReference.java:585) [jna-min-5.14.0.jar:5.14.0 (b0)]
        at com.sun.jna.CallbackReference$DefaultCallbackProxy.callback(CallbackReference.java:613) [jna-min-5.14.0.jar:5.14.0 (b0)]
        at com.sun.jna.Native.invokeVoid(Native Method) ~[jna-min-5.14.0.jar:5.14.0 (b0)]
        at com.sun.jna.Function.invoke(Function.java:418) ~[jna-min-5.14.0.jar:5.14.0 (b0)]
        at com.sun.jna.Function.invoke(Function.java:364) ~[jna-min-5.14.0.jar:5.14.0 (b0)]
        at com.sun.jna.Function.invoke(Function.java:318) ~[jna-min-5.14.0.jar:5.14.0 (b0)]
        at org.rococoa.internal.MsgSendHandler.invoke(MsgSendHandler.java:114) ~[rococoa-core-0.9.1.jar:?]
        at com.sun.jna.Library$Handler.invoke(Library.java:268) ~[jna-min-5.14.0.jar:5.14.0 (b0)]
        at jdk.proxy2/jdk.proxy2.$Proxy4.syntheticSendMessage(Unknown Source) ~[?:?]
        at org.rococoa.Foundation.send(Foundation.java:207) ~[rococoa-core-0.9.1.jar:?]
        at org.rococoa.Foundation.send(Foundation.java:181) ~[rococoa-core-0.9.1.jar:?]
        at org.rococoa.internal.ObjCObjectInvocationHandler.sendOnThisOrMainThread(ObjCObjectInvocationHandler.java:245) ~[rococoa-core-0.9.1.jar:?]
        at org.rococoa.internal.ObjCObjectInvocationHandler.invokeCocoa(ObjCObjectInvocationHandler.java:212) ~[rococoa-core-0.9.1.jar:?]
        at org.rococoa.internal.ObjCObjectInvocationHandler.intercept(ObjCObjectInvocationHandler.java:171) ~[rococoa-core-0.9.1.jar:?]
        at ch.cyberduck.binding.application.NSApplication$$ByRococoa.run(<generated>) ~[binding-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.ui.cocoa.MainApplication.main(MainApplication.java:71) ~[osx-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 1024 out of bounds for length 1024
        at com.jcraft.jsch.agentproxy.Buffer.getByte(Buffer.java:142) ~[jsch.agentproxy.core-0.0.10.jar:?]
        at com.jcraft.jsch.agentproxy.Buffer.getShort(Buffer.java:138) ~[jsch.agentproxy.core-0.0.10.jar:?]
        at com.jcraft.jsch.agentproxy.Buffer.getInt(Buffer.java:123) ~[jsch.agentproxy.core-0.0.10.jar:?]
        at com.jcraft.jsch.agentproxy.Buffer.getString(Buffer.java:181) ~[jsch.agentproxy.core-0.0.10.jar:?]
        at com.jcraft.jsch.agentproxy.AgentProxy.getIdentities(AgentProxy.java:114) ~[jsch.agentproxy.core-0.0.10.jar:?]
        at ch.cyberduck.core.sftp.openssh.OpenSSHAgentAuthenticator.getIdentities(OpenSSHAgentAuthenticator.java:62) ~[ssh-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.sftp.auth.SFTPAgentAuthentication.authenticate(SFTPAgentAuthentication.java:72) ~[ssh-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.sftp.auth.SFTPAgentAuthentication.authenticate(SFTPAgentAuthentication.java:58) ~[ssh-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.sftp.SFTPSession.authenticate(SFTPSession.java:346) ~[ssh-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.sftp.SFTPSession.login(SFTPSession.java:263) ~[ssh-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.KeychainLoginService.authenticate(KeychainLoginService.java:168) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.LoginConnectionService.authenticate(LoginConnectionService.java:163) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.LoginConnectionService.connect(LoginConnectionService.java:154) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.LoginConnectionService.check(LoginConnectionService.java:98) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.pool.StatelessSessionPool.borrow(StatelessSessionPool.java:56) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.threading.SessionBackgroundAction.run(SessionBackgroundAction.java:115) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.threading.SessionBackgroundAction$1.call(SessionBackgroundAction.java:106) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.threading.DefaultRetryCallable.call(DefaultRetryCallable.java:52) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.threading.SessionBackgroundAction.call(SessionBackgroundAction.java:108) ~[core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.threading.BackgroundCallable.run(BackgroundCallable.java:109) [core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at ch.cyberduck.core.threading.BackgroundCallable.call(BackgroundCallable.java:86) [core-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at java.base/java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
        at java.base/java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
        at ch.cyberduck.core.threading.DispatchExecutorService$InvocationFutureTask.run(DispatchExecutorService.java:281) [libcore-9.1.4-SNAPSHOT.jar:3b569cceae7546991bdd34f59b1b7aa80a143f3f]
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source) ~[?:?]
        at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[?:?]
        at org.rococoa.internal.OCInvocationCallbacks.callMethod(OCInvocationCallbacks.java:170) [rococoa-core-0.9.1.jar:?]
        at org.rococoa.internal.OCInvocationCallbacks.access$200(OCInvocationCallbacks.java:52) [rococoa-core-0.9.1.jar:?]
        at org.rococoa.internal.OCInvocationCallbacks$2.callback(OCInvocationCallbacks.java:82) [rococoa-core-0.9.1.jar:?]
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source) ~[?:?]
        at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[?:?]
        at com.sun.jna.CallbackReference$DefaultCallbackProxy.invokeCallback(CallbackReference.java:585) ~[jna-min-5.14.0.jar:5.14.0 (b0)]
        at com.sun.jna.CallbackReference$DefaultCallbackProxy.callback(CallbackReference.java:613) ~[jna-min-5.14.0.jar:5.14.0 (b0)]

```

--- a/ssh/pom.xml
+++ b/ssh/pom.xml
@@ -24,7 +24,7 @@
     <packaging>jar</packaging>
     <properties>
         <sshj-version>0.38.0</sshj-version>
-        <jsch-agentproxy-version>0.0.10</jsch-agentproxy-version>
+        <jsch-agentproxy-version>0.0.11</jsch-agentproxy-version>
     </properties>
     <dependencies>
         <dependency>

--- a/ssh/src/main/java/ch/cyberduck/core/sftp/SFTPSession.java
+++ b/ssh/src/main/java/ch/cyberduck/core/sftp/SFTPSession.java
@@ -71,7 +71,6 @@ import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.stream.Collectors;
 
-import com.jcraft.jsch.agentproxy.AgentProxyException;
 import net.schmizz.concurrent.Promise;
 import net.schmizz.keepalive.KeepAlive;
 import net.schmizz.keepalive.KeepAliveProvider;
@@ -283,27 +282,19 @@ public class SFTPSession extends Session<SSHClient> {
         // Ordered list of preferred authentication methods
         final List<AuthenticationProvider<Boolean>> defaultMethods = new ArrayList<>();
         if(preferences.getBoolean("ssh.authentication.agent.enable")) {
-            final String identityAgent = new OpenSSHIdentityAgentConfigurator().getIdentityAgent(host.getHostname());
-            log.debug("Determined identity agent {} for {}", identityAgent, host.getHostname());
-            switch(Platform.getDefault()) {
-                case windows:
-                    defaultMethods.add(new SFTPAgentAuthentication(client, new PageantAuthenticator()));
-                    try {
-                        defaultMethods.add(new SFTPAgentAuthentication(client,
-                                new WindowsOpenSSHAgentAuthenticator(identityAgent)));
-                    }
-                    catch(AgentProxyException e) {
-                        log.warn("Agent proxy failed with {}", e.getMessage());
-                    }
-                    break;
-                default:
-                    try {
+            final String configuration = new OpenSSHIdentityAgentConfigurator().getIdentityAgent(host.getHostname());
+            if(configuration != null) {
+                final String identityAgent = LocalFactory.get(configuration).getAbsolute();
+                log.debug("Determined identity agent {} for {}", identityAgent, host.getHostname());
+                switch(Platform.getDefault()) {
+                    case windows:
+                        defaultMethods.add(new SFTPAgentAuthentication(client, new PageantAuthenticator()));
+                        defaultMethods.add(new SFTPAgentAuthentication(client, new WindowsOpenSSHAgentAuthenticator(identityAgent)));
+                        break;
+                    default:
                         defaultMethods.add(new SFTPAgentAuthentication(client, new OpenSSHAgentAuthenticator(identityAgent)));
-                    }
-                    catch(AgentProxyException e) {
-                        log.warn("Agent proxy failed with {}", e.getMessage());
-                    }
-                    break;
+                        break;
+                }
             }
         }
         defaultMethods.add(new SFTPPublicKeyAuthentication(client));

--- a/ssh/src/main/java/ch/cyberduck/core/sftp/auth/SFTPPublicKeyAuthentication.java
+++ b/ssh/src/main/java/ch/cyberduck/core/sftp/auth/SFTPPublicKeyAuthentication.java
@@ -24,8 +24,8 @@ import ch.cyberduck.core.LocaleFactory;
 import ch.cyberduck.core.LoginCallback;
 import ch.cyberduck.core.LoginOptions;
 import ch.cyberduck.core.exception.BackgroundException;
-import ch.cyberduck.core.exception.InteroperabilityException;
 import ch.cyberduck.core.exception.LoginCanceledException;
+import ch.cyberduck.core.exception.LoginFailureException;
 import ch.cyberduck.core.sftp.SFTPExceptionMappingService;
 import ch.cyberduck.core.threading.CancelCallback;
 
@@ -106,7 +106,8 @@ public class SFTPPublicKeyAuthentication implements AuthenticationProvider<Boole
                         pubKey = null;
                         break;
                     default:
-                        throw new InteroperabilityException(String.format("Unknown key format for file %s", privKey.getName()));
+                        log.warn("Unknown key format for file {}", privKey.getName());
+                        throw new LoginFailureException(String.format("Unknown key format for file %s", privKey.getName()));
                 }
                 provider.init(new InputStreamReader(privKey.getInputStream(), StandardCharsets.UTF_8),
                         pubKey != null ? new InputStreamReader(pubKey.getInputStream(), StandardCharsets.UTF_8) : null,

--- a/ssh/src/main/java/ch/cyberduck/core/sftp/openssh/OpenSSHAgentAuthenticator.java
+++ b/ssh/src/main/java/ch/cyberduck/core/sftp/openssh/OpenSSHAgentAuthenticator.java
@@ -29,7 +29,6 @@ import java.util.Collections;
 import java.util.List;
 
 import com.jcraft.jsch.agentproxy.AgentProxy;
-import com.jcraft.jsch.agentproxy.AgentProxyException;
 import com.jcraft.jsch.agentproxy.Identity;
 import com.jcraft.jsch.agentproxy.connector.SSHAgentConnector;
 import com.jcraft.jsch.agentproxy.usocket.JNAUSocketFactory;
@@ -39,7 +38,7 @@ public class OpenSSHAgentAuthenticator extends AgentAuthenticator {
 
     private final AgentProxy proxy;
 
-    public OpenSSHAgentAuthenticator(final String socket) throws AgentProxyException {
+    public OpenSSHAgentAuthenticator(final String socket) {
         this(new AgentProxy(new SSHAgentConnector(new JNAUSocketFactory(), socket)));
     }
 

--- a/ssh/src/main/java/ch/cyberduck/core/sftp/openssh/WindowsOpenSSHAgentAuthenticator.java
+++ b/ssh/src/main/java/ch/cyberduck/core/sftp/openssh/WindowsOpenSSHAgentAuthenticator.java
@@ -20,14 +20,13 @@ import java.io.IOException;
 import java.io.RandomAccessFile;
 
 import com.jcraft.jsch.agentproxy.AgentProxy;
-import com.jcraft.jsch.agentproxy.AgentProxyException;
 import com.jcraft.jsch.agentproxy.USocketFactory;
 import com.jcraft.jsch.agentproxy.connector.SSHAgentConnector;
 
 public class WindowsOpenSSHAgentAuthenticator extends OpenSSHAgentAuthenticator {
     private final static String SSH_AGENT_PIPE = "\\\\.\\pipe\\openssh-ssh-agent";
 
-    public WindowsOpenSSHAgentAuthenticator(final String socketPath) throws AgentProxyException {
+    public WindowsOpenSSHAgentAuthenticator(final String socketPath) {
         super(new AgentProxy(new SSHAgentConnector(new RandomAccessFileSocketFactory(), fixupSocketPath(socketPath))));
     }
 

--- a/ssh/src/test/java/ch/cyberduck/core/sftp/openssh/OpenSSHAgentAuthenticatorTest.java
+++ b/ssh/src/test/java/ch/cyberduck/core/sftp/openssh/OpenSSHAgentAuthenticatorTest.java
@@ -28,18 +28,19 @@ import java.util.Collection;
 import com.jcraft.jsch.agentproxy.AgentProxyException;
 import com.jcraft.jsch.agentproxy.Identity;
 
+import static junit.framework.Assert.assertTrue;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assume.assumeTrue;
 
 public class OpenSSHAgentAuthenticatorTest {
 
-    @Test(expected = AgentProxyException.class)
+    @Test
     public void testGetIdentities() throws Exception {
         assumeTrue(Factory.Platform.getDefault().equals(Factory.Platform.Name.mac));
         final OpenSSHAgentAuthenticator authenticator = new OpenSSHAgentAuthenticator(StringUtils.EMPTY);
-        final Collection<Identity> identities = authenticator.getIdentities();
         assertNotNull(authenticator.getProxy());
-        assertFalse(identities.isEmpty());
+        final Collection<Identity> identities = authenticator.getIdentities();
+        assertTrue(identities.isEmpty());
     }
 }

